sycl/source/detail/global_handler.cpp

```diff
@@ -0,0 +1,137 @@
+//==--------- global_handler.cpp --- Global objects handler ----------------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include <CL/sycl/detail/spinlock.hpp>
+#include <detail/global_handler.hpp>
+#include <detail/platform_impl.hpp>
+#include <detail/plugin.hpp>
+#include <detail/program_manager/program_manager.hpp>
+#include <detail/scheduler/scheduler.hpp>
+
+#include <vector>
+
+__SYCL_INLINE_NAMESPACE(cl) {
+namespace sycl {
+namespace detail {
+GlobalHandler *SyclGlobalObjectsHandler;
+// SpinLock is chosen because, unlike std::mutex, it can be zero initialized,
+// which spares us from dealing with global constructor/destructor call order.
+SpinLock GlobalWritesAllowed;
+SpinLock ShutdownLock;
+
+GlobalHandler::GlobalHandler() = default;
+GlobalHandler::~GlobalHandler() = default;
+
+GlobalHandler &GlobalHandler::instance() {
+  if (!SyclGlobalObjectsHandler) {
```

Comment:
If you can rely on magic static, consider using local static variable:
```
GlobalHandler &GlobalHandler::instance() {
    static SyclGlobalObjectsHandler* SyclGlobalObjectsHandler = new GlobalHandler();
    return SyclGlobalObjectsHandler;
}
```
P.S. you might need a side effect in ctor of GlobalHandler to track that the initialization is done (for shutdown purposes) or wrap `new GlobalHandler()` into init function with the side effect.

Reply:
I'm not sure there's real risk here. I've restructured code according to Vlad's suggestion, so now it looks like this:
```
if (Object) {
  return *Object;
}

std::lock_guard<SpinLock> Lock{SL};
if (!Object) {
  Object = new ObjClass();
}

return *Object;
}
```

Let's say, two threads enter such a function. One thread has acquired lock and started writing to the pointer. Another thread is trying to read the value and gets `nullptr`. Then it needs to acquire lock, but by the time it reaches second condition, the object has been created. So, it'll just skip to return. Even though there's a race condition, I don't think it is harmful, as only one thread is allowed to write, and second read is also guarded with a lock. For other threads, there'll be no race condition, as they'll just read that variable. On the other hand, by the time shutdown is called, all other threads must finish their workloads, otherwise it will result into segfault anyway.

Reply:
It works because pointers are atomic on x86 hardware level. From C++11 perspective, it is UB.

Why you do not consider "magic static" approach? The synchronization is much easier and you do not need explicit locking and there is no issues with races. Your example will be:
```
ObjClass& getObject() {
    static ObjClass& obj = *new ObjClass();
    return obj;
}
```

Reply:
Done.